### DISCUSSION TITLE
Use of Emoji

### DISCUSSION THREAD

**Discussion by arnaudroques** [plantuml/plantuml#820] [General]
[Last commit](https://github.com/plantuml/plantuml/tree/master/src/net/sourceforge/plantuml/emojitwo) integrates few SVG from [EmojiTwo](https://github.com/EmojiTwo/emojitwo)

It's really an alpha version [which allows you to do this](https://www.plantuml.com/plantuml/umla/TOxDJiCm383lUmhVKPUomv2YgbeHvSO5s0EagRk4nV6KEEBdwKcHE4qxMlu-VpvaCKaAXSqGl4LeUdZa2JGSCm8lKAAtF2DqnxlxvttxqD_WHk8F__x9qEa_Jb-1NVO_g7TjFqGMXmdmoz0bO3Ls-xRLFPWuhvrrkcs_EKvYYrGjamzOGmkpr7l_cmnYpXXXIKpWH2vPAtNouigqjKpAduryctXIkFxSM8T6ckmGHKs19qKc2oRrEWvFB-EMviq1urme_080). Only 4 emojis are available right now but we plan to add more (maybe all of them)

This way, we could use [standard unicode emoji](https://unicode.org/emoji/charts/full-emoji-list.html) without having font issues. 
We plan also to [use shortcuts defined here](https://github.com/ikatyang/emoji-cheat-sheet/blob/master/README.md).

Any thought ?

**Comment by arnaudroques** [comment]
You can now have:
 ![image](https://www.plantuml.com/plantuml/png/SoWkIImgAStDuNBCoKnELT2rKt3AJrAmKh2ngYpDjR9Jo2zIi54sGEQa5wMcva44k1nIyrA0cW40)
with the following syntax

```
@startuml
Alice -> Bob : <:yum:> or <#0:yum:> or <#green:yum:>
@enduml
```

We have integrated:

- [Unicode block 26](https://www.plantuml.com/plantuml/uml/SoWkIImgAStDuKhDpS_AL30out98pKi12W00)
- [Unicode block 1f3](https://www.plantuml.com/plantuml/uml/SoWkIImgAStDuKhDpS_AL31CC-PoICrB0Oe00000)
- [Unicode block 1f4](https://www.plantuml.com/plantuml/uml/SoWkIImgAStDuKhDpS_AL31CC-5oICrB0Oe00000)
- [Unicode block 1f5](https://www.plantuml.com/plantuml/uml/SoWkIImgAStDuKhDpS_AL31CC-LoICrB0Oe00000)
- [Unicode block 1f6](https://www.plantuml.com/plantuml/uml/SoWkIImgAStDuKhDpS_AL31CC-DoICrB0Oe00000)
- [Unicode block 1f9](https://www.plantuml.com/plantuml/uml/SoWkIImgAStDuKhDpS_AL31Ci-HoICrB0Oe00000)

Only some emojis are really useful.

**Reply by VladimirAlexiev** [reply]
Great! Improved the doc slightly: http://alphadoc.plantuml.com/doc/markdown/en/creole#2jq22b6bnqpkkwxxmwos

**Reply by The-Lum** [reply]
Hello @arnaudroques,

Here are some test:
>[![](http://www.plantuml.com/plantuml/svg/LOqn2y8m48Nt-nKddLl5HGWXTdFNmK4SQdDeyPf8abhyzoR5M5qEtllUnzK-D2uCFKFRE8qjdk1WUy90FdW1A8kLI9ErK6XTZ9l_sEG8PGcok3eYaok1knInypWICqvm4LSN7cWU7MylbArbwqG2AXsvd0gLt6TdcS8sUexqh7sKsAazwHyXaqTdxYgTBnlZ9sD3LLLmXfgC7df-0m00)](http://www.plantuml.com/plantuml/umla/LOqn2y8m48Nt-nKddLl5HGWXTdFNmK4SQdDeyPf8abhyzoR5M5qEtllUnzK-D2uCFKFRE8qjdk1WUy90FdW1A8kLI9ErK6XTZ9l_sEG8PGcok3eYaok1knInypWICqvm4LSN7cWU7MylbArbwqG2AXsvd0gLt6TdcS8sUexqh7sKsAazwHyXaqTdxYgTBnlZ9sD3LLLmXfgC7df-0m00)

And some question:
1. Why blue 🟦 is violet?  
1. Why red 🔴 is pink? 
_Whereas green 🟢 is well green... 😄_

**Reply by The-Lum** [reply]
_In fact that depends of the original emoji..._
Compare `swan` to `grapes`... :)
>[![](http://www.plantuml.com/plantuml/svg/SoWkIImgAStDuKfEB4fHI3PKYEOgBa_Ci-AoKJQm0hFi56oKpU5CA8DBLnUe8hqeDJKFeIeffpGLmIjAJO4fZeMQRWGqFRqeiI2r668-X0sr0S61Cb5iWOh1xO7oOJP1kM2x75BpKe0-0W00)](http://www.plantuml.com/plantuml/umla/SoWkIImgAStDuKfEB4fHI3PKYEOgBa_Ci-AoKJQm0hFi56oKpU5CA8DBLnUe8hqeDJKFeIeffpGLmIjAJO4fZeMQRWGqFRqeiI2r668-X0sr0S61Cb5iWOh1xO7oOJP1kM2x75BpKe0-0W00)

**Reply by arnaudroques** [reply]
Thanks for the report, we're going to improve this.
This is due to the current algorithm used. When you specify a color the an emoji, we first convert colors of the emoji to level of gray (from 0 to 255). The main issue comes from the fact that some emoji does not goes from 0 to 255. We will add some rule of three somewhere to get a full color.

**Reply by The-Lum** [reply]
Thanks @arnaudroques,

There are some others _today's_ tests, with `#100` or `#F00`:
>[![](http://www.plantuml.com/plantuml/svg/SoWkIImgAStDuKfEB4fHI3PIYEPABqeiI2ssuh9HDh22ikqKR9JDaJXm9fUkBfTDSdvAgXrGUNfHQceUifQadD9KP7vHQWf2fuq-UVeGzXaOeZZGmC0GcMjeO81i6IcQC6rAGzEKXinDGxM934snN0wfUId0Q000)](http://www.plantuml.com/plantuml/umla/SoWkIImgAStDuKfEB4fHI3PIYEPABqeiI2ssuh9HDh22ikqKR9JDaJXm9fUkBfTDSdvAgXrGUNfHQceUifQadD9KP7vHQWf2fuq-UVeGzXaOeZZGmC0GcMjeO81i6IcQC6rAGzEKXinDGxM934snN0wfUId0Q000)

In fact, the main question is: what are the color format of this field <#**xxx**:name:>?
Is-it the algorithm here:
https://github.com/plantuml/plantuml/blob/1d6d3e847752b30753c43190374c93caef4063d0/src/net/sourceforge/plantuml/creole/legacy/StripeSimple.java#L279

_But I am not expert with this java snippets..._

**Reply by arnaudroques** [reply]
> In fact, the main question is: what are the color format of this field <#**xxx**:name:>?

`#012` is a shorcut for  `#001122` (known as [short hex](http://www.websiteoptimization.com/speed/tweak/hex/))

We have created a **very short hex** on a single digit: `#4` which is a shortcut for `#444444` or `#444` which [should work everywhere](https://www.plantuml.com/plantuml/uml/SoWkIImgAStDuNBAJrBGjLDmpCbCJbMmKh19pi_9Bx9IBafDiSj8pSd9LyZCKm1oRFJ14dOA-KLmDOu85Ub5YPKm9LmEgNafGAC1).

Hope it helps!

**Reply by The-Lum** [reply]
Yes... (As I mentioned on https://plantuml.com/en/color)
Then: why `1` is is stronger than `F`?
Is it due to the wrong conversion...

**Reply by arnaudroques** [reply]
> stronger

Sorry, I am not sure to understand 'stronger' ?

`#1` means `#111111` which means almost black
`#F` means `#FFFFFF` which means white

Note that there are indeed some issues in emoji: this is under correction :-)

**Reply by The-Lum** [reply]
No Pb. That is clear. And this new notation  `#x` = `#xxxxxx` is clear.
Now... We wait your works on emoji, with level of gray, with color... 😃

**Reply by VladimirAlexiev** [reply]
@arnaudroques 
The doc days
"using short HEX value (3 digits): #RGB".

But I'd assume that is translated to `R0G0B0` not `RRGGBB`. Could you add a clarification that it's translated to  `RRGGBB`?

**Reply by arnaudroques** [reply]
Yes, it's translated to `RRGGBB`, so for example `ABC` means `AABBCC`.
We're going to add the clarification in the doc.

**Comment by arnaudroques** [comment]
So this is working [better with last snapshot](https://www.plantuml.com/plantuml/uml/JPAnJiCm48PtFyMXpWsqIwGeCiqITXWO44D8JYN2ZIlR9UBjyRdd-1Rh-yztyQTJrCc7mONRnQXnS1EC1_XGVXsMHhMxfoQHXdPtA5WNx0hs6pAe_Jv-ukmGbzBrPMvOaiCfDt_oyyVu_DaDL_Ht0UQa84CjGoT3Bq96ASEreiEPXN8a9NQAIc_sWX3GXwW5v4LREx9MuLfm9xWlp2IDSgc15uemUrZH66X7QwnhKezEvuDYvstY_Vilzz65fgDzJ0LD7_gThS7mh7ZPUOn7Oas_uYG7-EOwBp-QZkrkZIyjDbHLHPivuJBHJ_CF).

Now I suspect that there will be issues will dark background... I let you test ! 😄

**Reply by The-Lum** [reply]
After [a first test](https://www.plantuml.com/plantuml/uml/JP4zRyCW48PtViLLcPCqMI9PYCQBkxT3Xog3Fqwk5GmHu5fHrVzUe5nWEJtlVN2lZhDrhN7BB1ctxYvHCAErWny6NTjVHwCNDVHQQWETz0bWluplKoVhMpD0Vu1FPjTMLOnldgj80lZcaF6KiSxOF301ssxz4wD1LBchamjcPN2WvajQV_JhHzFUqFuRI1mjbE9KYheKJI48IpEK8pj9aw4ySg6JU4llUaPmQ9st1S5NEFGnsSfyAhWkkCbCL3YYL0MlGS9aOKKfWSUlgcAFYF69qGbfukFhJdsX84BW-vWGuQ7lIKjqBomTcyRIYC-9DnoA0QhS9dKL8Jnggz-ajDljTluoPrJ3Cii_), it seems good.
Thanks a lot. :+1:

**Comment by The-Lum** [comment]
Hello PlantUML team,

Here are some questions about this new functionality:
- What were the arguments for choosing '[EmojiTwo](https://github.com/EmojiTwo/emojitwo)' or '[twemoji.twitter](https://twemoji.twitter.com)'?
- Can you tell us what motivated this choice?
- Would it be possible to include, _in the future_, other pack?
e.g. [openmoji](https://github.com/hfg-gmuend/openmoji), _[or [noto](https://github.com/googlefonts/noto-emoji) (perhaps too Google!...)]_, ...

_FYI, I just discovered [openmoji](https://github.com/hfg-gmuend/openmoji) and [openmoji.org](https://openmoji.org)..._

**Reply by huyz** [reply]
> * Would it be possible to include, _in the future_, other pack?
>   e.g. [openmoji](https://github.com/hfg-gmuend/openmoji), _[or [noto](https://github.com/googlefonts/noto-emoji) (perhaps too Google!...)]_, ...

Can you explain the pros and cons?

I'm confused between all the different sets of emojis...

**Reply by The-Lum** [reply]
Hello @huyz, and all,

>Can you explain the pros and cons?
I'm confused between all the different sets of emojis...

Before pros and cons _(because it so personal or subjective case)_, perhaps we can search some elements of comparison...
For example:
- License
- Style or design, bright colour or pastel, flat or not, ...
- Number of icons

_Then, perhaps we can create a new page on the wiki about that._

See also this pseudo-similar request:
- #965

Regards.

**Reply by The-Lum** [reply]
> Then, perhaps we can create a new page on the wiki about that.

Then, see this new page:
- https://github.com/plantuml/plantuml/wiki/Comparison-of-Emoji-or-SVG-icon-pack-on-PlantUML

**Reply by VladimirAlexiev** [reply]
@The-Lum nice! Could you add a table comparing a few known emoji (e.g smiley, thumbs up, page, cogwheel, rocket) between the different icon sets?

Then I can add to documentation

 Thanks!

**Reply by The-Lum** [reply]
@VladimirAlexiev: Good request... 
Here is a first add, when the set is included on PlantUML...
